File: Leet-046. Permutations/solution-recursive-n!.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        
        if (nums == null || nums.length == 0) {
            return res;
        }
        
        dfs(nums, res, 0);
        
        return res;
    }
    
    public void dfs(int[] nums, List<List<Integer>> res, int swapIdx) {
        if (swapIdx == nums.length) {
            List<Integer> list = new ArrayList<>();
            for (int val : nums) {
                list.add(val);
            }
            
            res.add(list);
            return ;
        }
        
        for (int i = swapIdx; i < nums.length; i++) {
            swap(nums, swapIdx, i);
            dfs(nums, res, swapIdx + 1);
            swap(nums, swapIdx, i);
        }
    }
    
    public void swap(int[] nums, int left, int right) {
        int temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
    }
}

File: Leet-046. Permutations/solution-recursive-n!.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    if (nums === null || nums.length < 2) {
        return [nums];
    }
    
    var results = [];
    backtrack(nums, results, 0);
    
    return results;
};

var backtrack = function(nums, results, swapIdx) {
    if (swapIdx === nums.length) {
        results.push(nums.slice());
    }
    
    for (var i = swapIdx; i < nums.length; i++) {
        swap(nums, i, swapIdx);
        backtrack(nums, results, swapIdx + 1);
        swap(nums, i, swapIdx);
    }
};

var swap = function(nums, start, end) {
    var temp = nums[start];
    nums[start] = nums[end];
    nums[end] = temp;
};